parser/src/transforms/inlining.rs

```diff
@@ -431,7 +447,50 @@ impl<'a> Inlining<'a> {
                 other => unimplemented!("unhandled builtin: {}", other),
             }
         } else {
-            todo!("pure functions are not implemented yet")
+            self.expand_function_callsite(call)
+        }
+    }
+
+    fn expand_binary_expr(&mut self, expr: &mut BinaryExpr) -> Result<Vec<Statement>, SemanticAnalysisError> {
+        let mut statements = self.expand_scalar_expr(&mut expr.lhs)?;
+        let rhs = self.expand_scalar_expr(&mut expr.rhs)?;
+        statements.extend(rhs);
+        
+        Ok(statements)
+    }
+
+    fn expand_scalar_expr(&mut self, expr: &mut ScalarExpr) -> Result<Vec<Statement>, SemanticAnalysisError> {
+        match expr {
+            ScalarExpr::Call(call) => {
+                let mut expanded = self.expand_call(call.clone())?;
+                with_let_result(self, &mut expanded, move |_, value| {
+                    *expr = ScalarExpr::try_from(value.clone())?;
+                    Ok(None)
+                })?;
+                Ok(expanded)
+            }
+            ScalarExpr::Binary(BinaryExpr {
+                op,
+                ref mut lhs,
+                ref mut rhs,
+                span,
+            }) => {
+                let mut stmts = Vec::new();
+
+                // Recursively expand the lhs and rhs.
+                stmts.extend(self.expand_scalar_expr(lhs)?);
```

Comment:
This is an example of a place where we need to be careful about how we handle the results of `expand_scalar_expr`. In particular, we need to handle this similarly to what I described in `expand_binary_expr`. Expand both operands, and then do the following:

Check the expansion of the left-hand operand to see if it terminates with a `let`. If it does, we're going to use `with_let_result` on it to perform a transformation like we did in `expand_binary_expr`. In the call to `with_let_result`, we will  check if the right-hand expansion also terminates with a let, and if so, we're going to immediately call `with_let_result` on it, and emit the binary operator in the body of a new `let`. The, now rewritten, expansion of the right-hand operand will be used as the body of a new `let` in the expansion of the left-hand statement block.

At the end, we simply return the expanded left-hand statement block (as we've moved the right-hand statement block into it).

NOTE: You won't need to call `rewrite_scalar_expr`, as by expanding both operands, that function gets called already.

NOTE: If either operand expands to a block that doesn't terminate in a `let`, which should only happen if there is just a single `Statement::Expr` in the block, you can emit the `with_let_result` for that operand, and just use that expression as the operand for the binary operator directly. Obviously, in the case where both operands are simple expressions, you can omit generating a `let` entirely, just as in `expand_binary_expr`